src/main/java/com/google/sps/model/GmailClientImpl.java

```diff
@@ -137,16 +187,57 @@ public Message getUserMessageWithMetadataHeaders(String messageId, List<String>
    */
   private List<Message> listUserMessagesWithMetadataHeaders(
       String searchQuery, List<String> metadataHeaders) throws IOException {
-    return listUserMessages(searchQuery).stream()
-        .map(
-            (message) -> {
-              try {
-                return getUserMessageWithMetadataHeaders(message.getId(), metadataHeaders);
-              } catch (IOException e) {
-                throw new RuntimeException(e);
-              }
-            })
-        .collect(Collectors.toList());
+    List<Message> userMessagesWithoutFormat = listUserMessages(searchQuery);
+    List<Message> userMessagesWithFormat = new ArrayList<>();
+
+    // When each message is retrieved, add them to the userMessagesWithFormat list
+    JsonBatchCallback<Message> batchCallback = addToListMessageCallback(userMessagesWithFormat);
+
+    // Add messages to a batch request, BATCH_REQUEST_CALL_LIMIT messages at a time
+    // At time of writing, the limit is 100 messages, so it will add 100 messages per request
+    int messageIndex = 0;
+    while (messageIndex < userMessagesWithoutFormat.size()) {
+      BatchRequest batchRequest = gmailService.batch();
+
+      while (messageIndex < userMessagesWithoutFormat.size()
+          && batchRequest.size() < BATCH_REQUEST_CALL_LIMIT) {
+        Message currentMessage = userMessagesWithoutFormat.get(messageIndex);
+        gmailService
+            .users()
+            .messages()
+            .get("me", currentMessage.getId())
+            .setFormat(MessageFormat.METADATA.formatValue)
+            .setMetadataHeaders(metadataHeaders)
+            .queue(batchRequest, batchCallback);
+        messageIndex++;
+      }
+
+      batchRequest.execute();
+    }
```

Comment:
At a glance this looks the same as the loop in `listUserMessagesWithFormat`. Would it be possible to abstract this out?

Reply:
Yeah I think I figured this out!